include/osv/sched.hh

```diff
@@ -313,6 +313,24 @@ private:
     int _renormalize_count;
 };
 
+// "Normal" threads are fairly time-shared according to the thread_runtime
+// data above. Such normal threads all have _realtime._priority == 0.
+// A "real-time" thread is one where _realtime._priority > 0. The scheduler
+// always picks the thread with the highest _realtime_priority to run next;
+// In particular normal threads run only when no real-time thread wants to
+// run. When several real-time threads with equal _realtime._priority want to
+// run, each one is run for _realtime._time_slice before switching to the
+// next one; If _realtime_time_slice is == 0, there is no limit on the amount
+// of time the thread may run - it will only be preempted when it waits,
+// yields, or a higher-priority thread comes along.
```

Comment:
I think this comment helps me answer some of my other questions above and I understand how the real-time thread may be preempted by yield (the `called_from_yield` related ifs) or if a higher-priority thread comes along. But I am not sure how it would be preempted when it waits - if it has the highest priority it would stay running until it changes or yields, no?

Reply:
If the realtime-priority thread **waits**, it goes out of the run queue, and not considered for running. Only later, when that thread is done waiting, it is re-inserted into the run queue. At that point it will preempt anything with a lower priority, but if something is running with exactly the same realtime priority, it will be put - as usual (enqueue()) - last among its equals - the thread running at that point will continue to run.